interfaces: interface hooks implementation

stolowski opened the pull request:
This PR integrates all the changes made so far and makes interface hooks work end-to-end, including auto-connections. However the support for snap refresh via new "reconnect" task will be proposed in a separate PR due to the need for some extra conflict-checking code.

bboozzoo commented on interfaces/repo.go:
```diff
@@ -539,9 +539,19 @@ func (r *Repository) ResolveDisconnect(plugSnapName, plugName, slotSnapName, slo
 // Connect establishes a connection between a plug and a slot.
 // The plug and the slot must have the same interface.
-func (r *Repository) Connect(ref ConnRef) error {
+func (r *Repository) Connect(ref ConnRef, plugAttrs map[string]interface{}, slotAttrs map[string]interface{}, policyCheck func(*ConnectedPlug, *ConnectedSlot) error) (*Connection, error) {
```
Not a problem, but it would be nice to pack the extra arguments into a structure, in current form it's quite unreadable

```
type ValidationData struct {
    PlugAttrs map[string]interface{}
    SlotAttrs map[string]interface{}
    PolicyChecker func(*ConnectedPlug, *ConnectedSlot
}
..

_, err = repo.Connect(interfaces.ConnRef{
		PlugRef: interfaces.PlugRef{Snap: info11.Name(), Name: ifname},
 		SlotRef: interfaces.SlotRef{Snap: core11.Name(), Name: ifname},
}, ValidationData{})
```

niemeyer commented on overlord/hookstate/ctlcmd/get.go:
```diff
@@ -291,15 +291,38 @@ func (c *getCommand) getInterfaceSetting(context *hookstate.Context, plugOrSlot
+		}
+
+		var value interface{}
+		err = config.GetFromChange(context.SnapName(), subkeys, 0, attributes["static"].(map[string]interface{}), &value)
```
Instead of deep-nesting, we can have "plug-dynamic" and "plug-static" in place of "plug-attrs".

Also, the logic above seems dangerous. It's type-asserting attributes["static"].. if it's missing for whatever reason, this logic will panic due to the type assertion.

niemeyer commented on overlord/ifacestate/handlers.go:
```diff
@@ -336,6 +333,40 @@ func (m *InterfaceManager) undoDiscardConns(task *state.Task, _ *tomb.Tomb) erro
+	slotAttrs["dynamic"] = dynamicSlotAttrs
+
+	task.Set("plug-attrs", plugAttrs)
+	task.Set("slot-attrs", slotAttrs)
```
Similar note for these functions. "plug-static" and "plug-dynamic" would simplify them.

niemeyer commented on overlord/ifacestate/handlers.go:
```diff
@@ -346,7 +377,7 @@ func (m *InterfaceManager) doConnect(task *state.Task, _ *tomb.Tomb) error {
 	}
 
-	conns, err := getConns(st)
+	connections, err := getConns(st)
```
Change seems unwanted considering everywhere else continues to refer to this specific map as conns, including the function itself at the right-hand side and other touched areas in the patch below.

niemeyer commented on overlord/ifacestate/helpers.go:
```diff
@@ -251,8 +252,10 @@ func (m *InterfaceManager) removeSnapSecurity(task *state.Task, snapName string)
+	Auto             bool                   `json:"auto,omitempty"`
+	Interface        string                 `json:"interface,omitempty"`
+	DynamicPlugAttrs map[string]interface{} `json:"dynamic-plug-attrs,omitempty"`
+	DynamicSlotAttrs map[string]interface{} `json:"dynamic-slot-attrs,omitempty"`
```
These might be "plug-dynamic" and "slot-dynamic" as well, reusing the terms from the task.

niemeyer commented on overlord/ifacestate/ifacestate.go:
```diff
@@ -134,7 +137,10 @@ func setInitialConnectAttributes(ts *state.Task, plugSnap string, plugName strin
+		attrs := make(map[string]interface{})
+		attrs["static"] = plug.Attrs
+		attrs["dynamic"] = make(map[string]interface{})
+		ts.Set("plug-attrs", attrs)
```
Same as earlier points about plug-static and plug-dynamic,.

niemeyer commented on overlord/ifacestate/ifacestate.go:
```diff
@@ -148,7 +154,10 @@ func setInitialConnectAttributes(ts *state.Task, plugSnap string, plugName strin
+		attrs := make(map[string]interface{})
+		attrs["static"] = slot.Attrs
+		attrs["dynamic"] = make(map[string]interface{})
+		ts.Set("slot-attrs", attrs)
```
Same.

stolowski commented on interfaces/repo.go:
```diff
@@ -539,9 +539,19 @@ func (r *Repository) ResolveDisconnect(plugSnapName, plugName, slotSnapName, slo
 }
 
+// SlotValidator can be implemented by Interfaces that need to validate the slot before the security is lifted.
+type SlotValidator interface {
```
Okay, for now I just made it lowercase as it doesn't need to be exported.

stolowski commented on interfaces/connection.go:
```diff
@@ -136,6 +136,11 @@ func (plug *ConnectedPlug) StaticAttrs() map[string]interface{} {
 }
 
+// DynamicAttrs returns all dynamic attributes.
+func (plug *ConnectedPlug) DynamicAttrs() map[string]interface{} {
```
As discussed leaving this getter as it's needed in a couple of other contexts. Fixed it (along with StaticAttrs()) to return deep copies.

stolowski commented on interfaces/policy/policy.go:
```diff
@@ -152,19 +176,19 @@ func nestedGet(which string, attrs map[string]interface{}, path string) (interfa
 
 func (connc *ConnectCandidate) PlugAttr(arg string) (interface{}, error) {
-	return nestedGet("plug", connc.Plug.Attrs, arg)
+	return nestedGet("plug", connc.plugAttrs(), arg)
```
I've changed to code to temporarily keep merged attributes in the ConnectCandidate, and only do the merging once. This is marked with FIXMEs and will be replaced in a followup PR which refactors policy code slightly to use Attrer interface.

stolowski commented:
@niemeyer I've addressed all you 1st pass comments, it's ready for re-review.

zyga commented on tests/main/interfaces-hooks/task.yaml:
```diff
@@ -10,8 +11,32 @@ prepare: |
+    rm -f "$CONSUMER_DATA/connect-plug-consumer-done"
+    rm -f "$PRODUCER_DATA/connect-slot-producer-done"
+
+environment:
```
Nitpick, move this immediately after summary so that human readers will first read the environment section and then see the referenced variables in the restore section.

zyga commented on tests/main/interfaces-hooks/task.yaml:
```diff
@@ -10,8 +11,32 @@ prepare: |
+    [ -f "$PRODUCER_DATA/connect-slot-producer-done" ]
+
+    systemctl stop snapd.service snapd.socket
+
```
Can you comment on what is going on here please? (in the code)

zyga commented on tests/main/interfaces-hooks/task.yaml:
```diff
@@ -10,8 +11,32 @@ prepare: |
+    jq -r '.data["conns"]["basic-iface-hooks-consumer:consumer basic-iface-hooks-producer:producer"]["slot-static"]["producer-attr2"]' /var/lib/snapd/state.json | MATCH "producervalue2"
+    jq -r '.data["conns"]["basic-iface-hooks-consumer:consumer basic-iface-hooks-producer:producer"]["plug-dynamic"]["consumer-attr3"]' /var/lib/snapd/state.json | MATCH "consumervalue3-validated"
+    jq -r '.data["conns"]["basic-iface-hooks-consumer:consumer basic-iface-hooks-producer:producer"]["slot-dynamic"]["producer-attr3"]' /var/lib/snapd/state.json | MATCH "producervalue3-validated"
+    systemctl start snapd.service
```
This is probably not needed (restore could handle that)

zyga commented on interfaces/builtin/test_snapd.go:
```diff
@@ -0,0 +1,99 @@
+	}
+}
+
+func (iface *testSnapdInterface) BeforePreparePlug(plug *snap.PlugInfo) error {
```
Are those mandatory?

zyga commented on interfaces/builtin/test_snapd.go:
```diff
@@ -0,0 +1,99 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
```
What's the point of this interface? I see it referenced from one of the test snaps below. How is that expected to work given that it won't be a part of non-test builds (due to file name).

zyga commented on tests/lib/snaps/test-snapd-policy-app-consumer/meta/snap.yaml:
```diff
@@ -322,3 +325,5 @@ plugs:
     bus: system
     name: test.system
+  test-snapd:
+    interface: test-snapd
```
Hmm, we have a new interface called `test-snapd`?

stolowski replied:
Renaming to "dummy" as discussed on IRC (and I'm open to rename it again if we find a better name; also we might consider hiding the interface from the snap interfaces list if needed, in a followup PR potentially).

zyga replied:
Thanks, this sounds great

zyga commented on tests/lib/snaps/basic-iface-hooks-producer/meta/snap.yaml:
```diff
@@ -1,8 +1,7 @@
\ No newline at end of file
+    producer:
+        interface: test-snapd
+        producer-attr1: producervalue1
```
Could you please use `producer-value-1` (unless you feel strongly about it here), the key could be called `producer-attr-1`.

zyga commented on tests/lib/snaps/basic-iface-hooks-producer/meta/hooks/prepare-slot-producer:
```diff
@@ -0,0 +1,33 @@
+#!/bin/sh
+
```
Can you please add `set -ue`

zyga commented on tests/lib/snaps/basic-iface-hooks-producer/meta/hooks/connect-slot-producer:
```diff
@@ -0,0 +1,50 @@
+#!/bin/sh
+
```
Can you please add `set -ue`

zyga commented on tests/lib/snaps/basic-iface-hooks-consumer/meta/hooks/prepare-plug-consumer:
```diff
@@ -0,0 +1,33 @@
+#!/bin/sh
+
```
Can you please add `set -ue`

zyga commented on tests/lib/snaps/basic-iface-hooks-consumer/meta/hooks/connect-plug-consumer:
```diff
@@ -0,0 +1,60 @@
+#!/bin/sh
+
```
Can you please add `set -ue`

zyga commented on tests/lib/snaps/basic-iface-hooks-consumer/meta/hooks/connect-plug-consumer:
```diff
@@ -0,0 +1,60 @@
+#!/bin/sh
```
Can you shellcheck all of the new hooks please?

zyga commented on overlord/ifacestate/handlers.go:
```diff
@@ -428,9 +413,18 @@ func (m *InterfaceManager) doConnect(task *state.Task, _ *tomb.Tomb) error {
 	}
 
-	conns[connRef.ID()] = connState{Interface: plug.Interface}
+	conns[connRef.ID()] = connState{
```
If we are storing static attributes this may misbehave when the snap is changed on disk (snap try will be affected by this). How do you plan to handle that?

stolowski replied:
Thanks for pointing that out. I need to consider this, seems like a valid concern.

zyga replied:
So this is my last concern for this PR.

I think the question goes back to should snapd cache part of snap info from disk and how it should behave when things can change under us (especially in snap try situations)

zyga commented on overlord/ifacestate/handlers.go:
```diff
@@ -336,6 +333,26 @@ func (m *InterfaceManager) undoDiscardConns(task *state.Task, _ *tomb.Tomb) erro
+	var plugAttrs, slotAttrs map[string]interface{}
+
+	var err error
+	if err = task.Get("plug-dynamic", &plugAttrs); err != nil {
```
Nitpick: You can drop `var err error` above and just `:=` assign the err variable here.

zyga commented on overlord/hookstate/ctlcmd/set.go:
```diff
@@ -132,18 +156,23 @@ func (s *setCommand) setInterfaceSetting(context *hookstate.Context, plugOrSlot
+		return fmt.Errorf(i18n.G("internal error: cannot get %s from appropriate task, %s"), which, err)
+	}
+
+	dynKey := fmt.Sprintf("%s-dynamic", which)
```
Nitpick: can you roll the `dynKey` variable into the `Get` call below as you did above?

stolowski replied:
I need it again a few lines below, that's why it goes into dynKey.

zyga commented on overlord/hookstate/ctlcmd/set.go:
```diff
@@ -113,6 +115,28 @@ func (s *setCommand) setConfigSetting(context *hookstate.Context) error {
+	}
+
+	var existing interface{}
+	err = config.GetFromChange(context.SnapName(), subkeys[:1], 0, staticAttrs, &existing)
```
What guarantees that `subkeys` is not empty?

zyga replied:
Can you add a comment with that text please?

zyga commented on overlord/hookstate/ctlcmd/get.go:
```diff
@@ -282,24 +282,42 @@ func (c *getCommand) getInterfaceSetting(context *hookstate.Context, plugOrSlot
-	if err = attrsTask.Get(which, &attributes); err != nil {
+	staticAttrs := make(map[string]interface{})
+	dynamicAttrs := make(map[string]interface{})
+	if err = attrsTask.Get(fmt.Sprintf("%s-static", which), &staticAttrs); err != nil {
```
Do you have to `make` those? Isn't `Get` going to do that for you? The same question applies to similar code later in this diff.

stolowski replied:
Thanks for spotting, not needed indeed. Fixed.

zyga commented on interfaces/repo.go:
```diff
@@ -970,7 +1003,7 @@ func (r *Repository) AutoConnectCandidatePlugs(slotSnapName, slotName string, po
 			slot := &Slot{SlotInfo: slotInfo}
 			// declaration based checks disallow
-			if !policyCheck(plug, slot) {
+			if !policyCheck(NewConnectedPlug(plugInfo, nil), NewConnectedSlot(slotInfo, nil)) {
```
Should we pay attention to making sure that connected plug / slot objects are only created in a controlled way and kept somewhere in the repository/daemon/wherever (so that there's one such object per connection).

zyga commented on interfaces/repo.go:
```diff
@@ -935,7 +968,7 @@ func (r *Repository) AutoConnectCandidateSlots(plugSnapName, plugName string, po
 			slot := &Slot{SlotInfo: slotInfo}
 			// declaration based checks disallow
-			if !policyCheck(plug, slot) {
+			if !policyCheck(NewConnectedPlug(plugInfo, nil), NewConnectedSlot(slotInfo, nil)) {
```
Same question as below (sorry, I'm reading the diff from the end).

zyga commented on interfaces/repo.go:
```diff
@@ -553,23 +563,50 @@ func (r *Repository) Connect(ref ConnRef) error {
+	cplug := NewConnectedPlug(plug, plugDynamicAttrs)
+	cslot := NewConnectedSlot(slot, slotDynamicAttrs)
+
+	iface, ok := r.ifaces[plug.Interface]
```
Can you move this check above, so that it happens before we construct `NewConnected{Plug,Slot}` please?

stolowski commented:
The branch is ready for reviews again. Note that the sub-optimal merging of dynamic and static attributes when passing them to policy checks is dealt with in the separate PR.

niemeyer commented on interfaces/builtin/dummy.go:
```diff
@@ -0,0 +1,99 @@
+	return "dummy"
+}
+
+func (iface *testSnapdInterface) StaticInfo() interfaces.StaticInfo {
```
Type name needs to match the interface name as usual.

stolowski replied:
Done.

niemeyer commented on interfaces/builtin/dummy.go:
```diff
@@ -0,0 +1,99 @@
+
+func (iface *testSnapdInterface) BeforeConnectPlug(plug *interfaces.ConnectedPlug) error {
+	var value string
+	if err := plug.Attr("consumer-attr-3", &value); err != nil {
```
Even being a test interface, ideally its attributes would look nice and give a hint of their meaning.

How about renaming both of those attributes to _"before-connect"_ (both the same), and then doing the value replacement like this: 
```
value = fmt.Sprintf("plug-changed(%v)", value)
```
and the equivalent for the slot.

Besides looking nice, this makes sure that we handle matching names correctly (while we don't generally want that in real interfaces, each of these attributes needs to live in their own namespace), and we also make sure that the logic was provided to the right end of it (with "foo-validated" they are both doing the same change).

stolowski replied:
Thanks, that's a good suggestion. done.

niemeyer commented on interfaces/builtin/dummy.go:
```diff
@@ -0,0 +1,99 @@
+
+type testSnapdInterface struct{}
+
+const testSnapdInterfaceSummary = `snapd dummy test interface`
```
That's not following our usual pattern for these summaries. Also, if it's _"dummy",_ it needs to be more general than just snapd itself.

Perhaps something like this:

_"allows testing without providing any additional permissions"_

stolowski replied:
Done.

pedronis commented on overlord/ifacestate/ifacestate.go:
```diff
@@ -203,8 +210,11 @@ func setInitialConnectAttributes(ts *state.Task, plugSnap string, plugName strin
+	emptyDynamicAttrs := make(map[string]interface{})
 	if plug, ok := snapInfo.Plugs[plugName]; ok {
-		ts.Set("plug-attrs", plug.Attrs)
+		ts.Set("plug-static", plug.Attrs)
```
what happens on update with old tasks that had the previous thing?

stolowski replied:
Ok, I thought about that and we're good. We're only reading static attributes in snapctl from interface hooks (but no one should be using them yet). And when it comes to dynamic attributes, with the change to getDynamicHookAttributes you suggested below we will ignore ErrNoState and use empty dynamic attributes. ets

pedronis commented on overlord/ifacestate/ifacestate.go:
```diff
@@ -187,6 +189,11 @@ func connect(st *state.State, change *state.Change, plugSnap, plugName, slotSnap
 
+	// store IDs of the connect-plug/slot- hook tasks in the main 'connect' task,
+	// so that doConnect can conditionally mark them done on autoconnect.
+	connectInterface.Set("connect-slot-task", connectSlotConnection.ID())
```
should this name contain hook?

stolowski replied:
Done.

pedronis commented on overlord/ifacestate/ifacestate.go:
```diff
@@ -187,6 +189,11 @@ func connect(st *state.State, change *state.Change, plugSnap, plugName, slotSnap
+	// store IDs of the connect-plug/slot- hook tasks in the main 'connect' task,
+	// so that doConnect can conditionally mark them done on autoconnect.
+	connectInterface.Set("connect-slot-task", connectSlotConnection.ID())
+	connectInterface.Set("connect-plug-task", connectPlugConnection.ID())
```
same

pedronis commented on overlord/ifacestate/handlers.go:
```diff
@@ -328,6 +325,41 @@ func (m *InterfaceManager) undoDiscardConns(task *state.Task, _ *tomb.Tomb) erro
+	var plugAttrs, slotAttrs map[string]interface{}
+
+	if err := task.Get("plug-dynamic", &plugAttrs); err != nil {
+		return nil, nil, err
```
should we ignore NoState errors to be more backward compatible?

stolowski replied:
Done.

pedronis commented on interfaces/repo.go:
```diff
@@ -673,6 +706,34 @@ func (r *Repository) connected(snapName, plugOrSlotName string) ([]ConnRef, erro
 	return conns, nil
 }
 
+func (r *Repository) Connections(snapName string) ([]ConnRef, error) {
```
I would have expected []*ConnRef

stolowski replied:
This is in line with other methods in repo that return ConnRef (ResolveDisconnect, Connected).

pedronis replied:
well, we are really not that consistent with xxxRef  types,  I see

NewConnRef(plug *snap.PlugInfo, slot *snap.SlotInfo) *ConnRef
vs
func ParseConnRef(id string) (ConnRef, error)
vs
func (slot *Slot) Ref() SlotRef
vs
func (slot *ConnectedSlot) Ref() *SlotRef

not here, but for a follow up, unless we compare these xxxRef a lot with each other or use them as keys in maps a lot,  I would use  *xxxRef everywhere,  they are big enough (>=128bits) for that to make sense

also seeing *NewConnRef(...) a lot is strange

stolowski replied:
Ok, will do in a followup.

stolowski replied:
Fair enough; I made the change in separate PR where we can discuss the merit.

pedronis commented on interfaces/policy/policy.go:
```diff
@@ -116,22 +117,49 @@ func (ic *InstallCandidate) Check() error {
-	return connc.Plug.Attrs
+	// FIXME: change policy code to use Attrer interface, remove merging.
+	if connc.PlugAttrs == nil {
+		connc.PlugAttrs = mergedAttributes(connc.Plug.StaticAttrs(), connc.Plug.DynamicAttrs())
```
do we have tests for when DynamicAttrs is not empty?

I'm not spotting them in policy_test.go, but I might be missing them

stolowski replied:
You're right, they were missing. Added.

pedronis commented on interfaces/policy/policy.go:
```diff
@@ -116,22 +117,49 @@ func (ic *InstallCandidate) Check() error {
-	return connc.Slot.Attrs
+	// FIXME: change policy code to use Attrer interface, remove merging.
+	if connc.SlotAttrs == nil {
+		connc.SlotAttrs = mergedAttributes(connc.Slot.StaticAttrs(), connc.Slot.DynamicAttrs())
```
same question

stolowski replied:
Added.

pedronis commented on interfaces/policy/policy.go:
```diff
@@ -116,22 +117,49 @@ func (ic *InstallCandidate) Check() error {
-	Plug                *snap.PlugInfo
+	Plug *interfaces.ConnectedPlug
+	// Static + dynamic attributes, merged lazily by plugAttrs below (FIXME: remove it)
+	PlugAttrs           map[string]interface{}
```
does this need to be public given that is a cache and will go away?

stolowski replied:
Done.

pedronis commented on interfaces/policy/policy.go:
```diff
@@ -116,22 +117,49 @@ func (ic *InstallCandidate) Check() error {
-	Slot                *snap.SlotInfo
+	Slot *interfaces.ConnectedSlot
+	// Static + dynamic attributes, merged lazily by slotAttrs below (FIXME: remove it)
+	SlotAttrs           map[string]interface{}
```
same question

stolowski replied:
Done.

pedronis commented on overlord/ifacestate/handlers.go:
```diff
@@ -507,6 +530,83 @@ func (m *InterfaceManager) doDisconnect(task *state.Task, _ *tomb.Tomb) error {
+		connectts.AddAll(ts)
+	}
+
+	task.SetStatus(state.DoneStatus)
```
this should be move toward the end, is a bit strange here

stolowski replied:
Done.

pedronis commented on overlord/ifacestate/handlers.go:
```diff
@@ -507,6 +530,83 @@ func (m *InterfaceManager) doDisconnect(task *state.Task, _ *tomb.Tomb) error {
+
+	task.SetStatus(state.DoneStatus)
+
+	lanes := task.Lanes()
```
we have this kind of code to add tasks and consider lanes in a couple of places now,  is it worth making an helper?

stolowski replied:
That's right. If that's OK with you then I'd prefer to address this in a followup PR and fix all these occurences in same PR, it will be more visible what's happening.

pedronis replied:
it's fine in a follow up

pedronis commented on overlord/hookstate/ctlcmd/get.go:
```diff
@@ -282,24 +282,41 @@ func (c *getCommand) getInterfaceSetting(context *hookstate.Context, plugOrSlot
 		}
-		return nil, false, fmt.Errorf(i18n.G("unknown attribute %q"), key)
+		if config.IsNoOption(err) {
+			err = config.GetFromChange(context.SnapName(), subkeys, 0, dynamicAttrs, &value)
```
does this bit have tests?

stolowski replied:
It didn't have unit test (only spread test) - added now. I also refactored all these unit tests - these new cases are about "dyn-slot-attr" and "dyn-plug-attr".

pedronis commented on overlord/hookstate/ctlcmd/set.go:
```diff
@@ -113,6 +115,31 @@ func (s *setCommand) setConfigSetting(context *hookstate.Context) error {
+	// part of key=value argument and is guaranteed to be non-empty at this
+	// point.
+	err = config.GetFromChange(context.SnapName(), subkeys[:1], 0, staticAttrs, &existing)
+	if err == nil {
```
we need to return the error if not a NoOption, no?

stolowski replied:
Good point, done.

stolowski commented on overlord/ifacestate/ifacestate_test.go:
```diff
@@ -2300,8 +2421,12 @@ func (s *interfaceManagerSuite) TestConnectIgnoresMissingPlugSnapOnAutoConnect(c
 
 	task := chg.Tasks()[0]
-	c.Assert(task.Status(), Equals, state.DoneStatus)
+	c.Check(task.Status(), Equals, state.DoneStatus)
```
Right, unintended change, restored.

niemeyer commented on overlord/hookstate/ctlcmd/set.go:
```diff
@@ -132,18 +163,22 @@ func (s *setCommand) setInterfaceSetting(context *hookstate.Context, plugOrSlot
+	defer context.Unlock()
+
+	var staticAttrs, dynamicAttrs map[string]interface{}
+	if err = attrsTask.Get(fmt.Sprintf("%s-static", which), &staticAttrs); err != nil {
```
which+"-static"

niemeyer commented on overlord/hookstate/ctlcmd/get.go:
```diff
@@ -282,24 +282,41 @@ func (c *getCommand) getInterfaceSetting(context *hookstate.Context, plugOrSlot
-	attributes := make(map[string]interface{})
-	if err = attrsTask.Get(which, &attributes); err != nil {
+	var staticAttrs, dynamicAttrs map[string]interface{}
+	if err = attrsTask.Get(fmt.Sprintf("%s-static", which), &staticAttrs); err != nil {
```
which+"-static", and similar below.

niemeyer commented on overlord/hookstate/ctlcmd/get_test.go:
```diff
@@ -205,115 +237,112 @@ func (s *getAttrSuite) SetUpTest(c *C) {
+		args:   "get :aplug aattr",
+		stdout: "foo\n",
+	},
+	{
```
For that sort of table, this format saves both vertical and horizontal spacing, rendering the code more readable:
```
...
}{{
    args: ...
    stdout: ...
}, {
    args: ...
    stdout: ...
}, {
...
```

niemeyer commented on overlord/hookstate/ctlcmd/get_test.go:
```diff
@@ -205,115 +237,112 @@ func (s *getAttrSuite) SetUpTest(c *C) {
+		args:   "get :bslot battr",
+		stdout: "bar\n",
+	},
+	{
```
Same about formatting.

niemeyer commented on overlord/hookstate/ctlcmd/set.go:
```diff
@@ -132,18 +163,22 @@ func (s *setCommand) setInterfaceSetting(context *hookstate.Context, plugOrSlot
-	if err := attrsTask.Get(which, &attributes); err != nil {
-		return fmt.Errorf(i18n.G("internal error: cannot get %s from appropriate task"), which)
+	dynKey := fmt.Sprintf("%s-dynamic", which)
+	if err = attrsTask.Get(dynKey, &dynamicAttrs); err != nil {
```
Get(which+"-dynamic", ...)

niemeyer commented on overlord/ifacestate/handlers.go:
```diff
@@ -349,6 +346,47 @@ func (m *InterfaceManager) undoDiscardConns(task *state.Task, _ *tomb.Tomb) erro
 	return nil
 }
 
+func getDynamicHookAttributes(task *state.Task) (map[string]interface{}, map[string]interface{}, error) {
```
This needs result names. As it is we need to read the code to find out what the results mean.

Doing that also allows avoiding the type repetition ( as in _(a, b type)_)

niemeyer commented on overlord/ifacestate/handlers.go:
```diff
@@ -349,6 +346,47 @@ func (m *InterfaceManager) undoDiscardConns(task *state.Task, _ *tomb.Tomb) erro
+	return plugAttrs, slotAttrs, nil
+}
+
+func setDynamicHookAttributes(task *state.Task, dynamicPlugAttrs map[string]interface{}, dynamicSlotAttrs map[string]interface{}) {
```
The type repetition here may be avoided per above note.

niemeyer commented on overlord/ifacestate/handlers.go:
```diff
@@ -349,6 +346,47 @@ func (m *InterfaceManager) undoDiscardConns(task *state.Task, _ *tomb.Tomb) erro
+}
+
+func markConnectHooksDone(connectTask *state.Task) error {
+	for _, t := range []string{"connect-plug-hook-task", "connect-slot-hook-task"} {
```
This looks a bit suspect. How come we mark it done if it wasn't done? The connect task for a particular snap cannot possibly succeed if the snap isn't available right? I'm probably missing something.. can we please talk about it?

stolowski replied:
This is only for auto-connect. In one of the older reviews @pedronis sugested that we ignore missing snaps at this point when auto-connecting in case the snap is gone by the time we execute this task. This egde case poses problem for interface hooks though as they will inevitably error out if they don't find given snap in the state.

niemeyer commented on overlord/ifacestate/handlers.go:
```diff
@@ -471,9 +485,24 @@ func (m *InterfaceManager) doConnect(task *state.Task, _ *tomb.Tomb) error {
+	}
 	setConns(st, conns)
 
+	// the dynamic attributes might have been updated by interface's BeforeConnectPlug/Slot code,
```
s/by/by the/

stolowski replied:
Done.

niemeyer commented on overlord/ifacestate/handlers.go:
```diff
@@ -528,6 +557,83 @@ func (m *InterfaceManager) doDisconnect(task *state.Task, _ *tomb.Tomb) error {
+	}
+
+	connectts := state.NewTaskSet()
+	chg := task.Change()
```
The change is not used until way down there. Moving it closer makes it slightly more clear.

stolowski replied:
Done.

niemeyer commented on overlord/ifacestate/helpers.go:
```diff
@@ -205,15 +205,17 @@ func (m *InterfaceManager) reloadConnections(snapName string) ([]string, error)
 	}
 	affected := make(map[string]bool)
-	for id := range conns {
+	for id, cn := range conns {
```
conn?  That's what was used above.

stolowski replied:
Ok, changed.

niemeyer commented on overlord/ifacestate/ifacestate.go:
```diff
@@ -187,6 +189,11 @@ func connect(st *state.State, change *state.Change, plugSnap, plugName, slotSnap
+	// store IDs of the connect-plug/slot- hook tasks in the main 'connect' task,
+	// so that doConnect can conditionally mark them done on autoconnect.
+	connectInterface.Set("connect-slot-hook-task", connectSlotConnection.ID())
+	connectInterface.Set("connect-plug-hook-task", connectPlugConnection.ID())
```
connect-slot-task and connect-plug-task might be enough for clarity here, and slightly less verbose.

stolowski replied:
This is something I had originally and changed per @pedronis request. I think having *hook* in the name makes it clear what this is about, whereas connect-slot-task / connect-plug-task might get confused with the main connect task.
How about chaning these to "slot-hook-task" and "plug-hook-task"?

niemeyer replied:
It's okay, let's preserve the names as they are now then.

niemeyer commented on overlord/snapstate/snapstate.go:
```diff
@@ -203,6 +203,13 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 
+	// on refresh re-connect existing connections and run their interface hooks (if applicable)
+	if snapst.IsInstalled() {
+		reconnectTask := st.NewTask("reconnect", fmt.Sprintf(i18n.G("Re-connect the interfaces and re-run interface hooks for snap %q"), snapsup.Name()))
```
"Reconnect interfaces for snap %q".. hooks are part of interfaces (note s/the//).

stolowski replied:
Done.

stolowski commented on interfaces/repo.go:
```diff
@@ -553,23 +566,48 @@ func (r *Repository) Connect(ref ConnRef) error {
+	if policyCheck != nil {
+		if i, ok := iface.(plugValidator); ok {
+			if err := i.BeforeConnectPlug(cplug); err != nil {
+				return nil, fmt.Errorf("cannot connect plug %q from snap %q: %s", plug.Name, plug.Snap.Name(), err)
```
Done.

stolowski commented on interfaces/repo.go:
```diff
@@ -553,23 +566,48 @@ func (r *Repository) Connect(ref ConnRef) error {
+
+	iface, ok := r.ifaces[plug.Interface]
+	if !ok {
+		// This should never happen
```
Done.

stolowski commented on interfaces/policy/policy.go:
```diff
@@ -116,22 +117,49 @@ func (ic *InstallCandidate) Check() error {
-	return connc.Plug.Attrs
+	// FIXME: change policy code to use Attrer interface, remove merging.
+	if connc.mergedPlugAttrs == nil {
+		connc.mergedPlugAttrs = mergedAttributes(connc.Plug.StaticAttrs(), connc.Plug.DynamicAttrs())
```
This is a temporary connect candidate object, it goes away after poliy check is done.

stolowski commented on interfaces/policy/policy.go:
```diff
@@ -116,22 +117,49 @@ func (ic *InstallCandidate) Check() error {
+		if _, ok := merged[k]; ok {
+			// Safeguard. This should never happen as it's prevented
+			// when attributes are populated at higher levels.
+			panic(fmt.Sprintf("internal error: attempted to overwrite static attribute %q", k))
```
Fixed.

stolowski commented on interfaces/builtin/dummy.go:
```diff
@@ -0,0 +1,99 @@
+        - app
+    deny-auto-connection: true
+    deny-connection:
+      on-classic: false
```
Indeed we don't need the deny-connection.

stolowski commented:
@niemeyer Thanks for the review! I've addressed your comments, except for the question of what to do with connect-slot-hook-task / connect-plug-hook-task - see my reply.

stolowski commented on overlord/hookstate/ctlcmd/get.go:
```diff
@@ -282,24 +282,41 @@ func (c *getCommand) getInterfaceSetting(context *hookstate.Context, plugOrSlot
+			if err == nil {
+				return value, true, nil
+			}
+			return nil, false, fmt.Errorf(i18n.G("unknown attribute %q"), key)
```
We just return it (without any extra decoration) below with `return nil, false, err`.

pedronis replied:
no, what  Gustavo meant is that if the sencod GetFromChange about dynamicAttrs returns something that is *not* IsNotOption we should return it straight, which this code doesn't do (it returns "unknown attribute")

stolowski commented on interfaces/policy/policy.go:
```diff
@@ -152,19 +176,19 @@ func nestedGet(which string, attrs map[string]interface{}, path string) (interfa
 
 func (connc *ConnectCandidate) SlotAttr(arg string) (interface{}, error) {
-	return nestedGet("slot", connc.Slot.Attrs, arg)
+	return nestedGet("slot", connc.slotAttrs(), arg)
```
See the reply above.

pedronis commented on overlord/ifacestate/handlers.go:
```diff
@@ -528,6 +555,84 @@ func (m *InterfaceManager) doDisconnect(task *state.Task, _ *tomb.Tomb) error {
+
+	connectts := state.NewTaskSet()
+	for _, conn := range connections {
+		ts, err := Connect(st, conn.PlugRef.Snap, conn.PlugRef.Name, conn.SlotRef.Snap, conn.SlotRef.Name)
```
as discussed this needs to wait if the other snap is being operated on (pending link/unlink/setup-profiles) like we do for auto-connect

stolowski commented:
This is ready for re-review; I've pulled support for refresh ("reconnect") as discussed, it will be proposed in a followup PR.

pedronis commented:
thanks, lgtm, afaict comment were resolved

stolowski commented:
Merging as agreed on the standup after a pass over Gustavo's comments made by Samuele.
